File: workers/_metabolism.js
/**
 * Metabolic stasis — isomorphic (Workers + browser + Node).
 * hourly_cap = remaining / hours_until_renewal
 * spendable  = remaining - reserved_future_peaks
 * No 6th Cloudflare cron. Identity ≠ cargo.
 */
export const ALLOW = "ALLOW";
export const HOLD = "HOLD";
export const STASIS = "STASIS";
export const P0_BORROW = "P0_BORROW";

const DEFAULT_TZ = "Europe/Lisbon";

export function parseHhmm(hhmm) {
  const [h, m] = String(hhmm).split(":").map((n) => parseInt(n, 10));
  return { h, m };
}

export function zonedParts(date, timeZone) {
  const fmt = new Intl.DateTimeFormat("en-GB", {
    timeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hourCycle: "h23",
  });
  const parts = Object.fromEntries(fmt.formatToParts(date).map((p) => [p.type, p.value]));
  return {
    year: +parts.year,
    month: +parts.month,
    day: +parts.day,
    hour: +parts.hour,
    minute: +parts.minute,
    second: +parts.second,
  };
}

function pad(n) {
  return String(n).padStart(2, "0");
}

/** Approximate timezone offset by comparing UTC ms of local wall time. */
function tzOffsetMs(wall, timeZone) {
  const utcGuess = Date.UTC(wall.year, wall.month - 1, wall.day, wall.hour, wall.minute, wall.second);
  const asTz = zonedParts(new Date(utcGuess), timeZone);
  const asTzUtc = Date.UTC(asTz.year, asTz.month - 1, asTz.day, asTz.hour, asTz.minute, asTz.second);
  return utcGuess - asTzUtc;
}

export function hoursUntilRenewal(now, renewalHHmm = "00:00", timeZone = DEFAULT_TZ) {
  const local = zonedParts(now, timeZone);
  const { h, m } = parseHhmm(renewalHHmm);
  let cand = { ...local, hour: h, minute: m, second: 0 };
  const localMin = local.hour * 60 + local.minute;
  const candMin = h * 60 + m;
  if (candMin <= localMin) {
    const dt = new Date(Date.UTC(local.year, local.month - 1, local.day) + 86400000);
    const next = zonedParts(dt, timeZone);
    cand = { year: next.year, month: next.month, day: next.day, hour: h, minute: m, second: 0 };
  }
  const nowMs = now.getTime();
  const candUtc = Date.UTC(cand.year, cand.month - 1, cand.day, cand.hour, cand.minute, 0) - tzOffsetMs(cand, timeZone);
  const hours = (candUtc - nowMs) / 3600000;
  return Math.max(hours, 1 / 60);
}

export function hoursUntilUnix(now, resetUnix) {
  return Math.max((resetUnix * 1000 - now.getTime()) / 3600000, 1 / 60);
}

function minutesOf(parts) {
  return parts.hour * 60 + parts.minute;
}

function hhmmMinutes(hhmm) {
  const { h, m } = parseHhmm(hhmm);
  return h * 60 + m;
}

export function isNight(now, spec = {}, timeZone = DEFAULT_TZ) {
  const tz = spec.renewal_tz || timeZone;
  const parts = zonedParts(now, tz);
  const start = hhmmMinutes(spec.night_start || "22:00");
  const end = hhmmMinutes(spec.night_end || "08:00");
  const m = minutesOf(parts);
  if (start > end) return m >= start || m < end;
  return m >= start && m < end;
}

export function slotsAhead(cfg, now, rail, excludeId) {
  const tz = cfg.timezone || DEFAULT_TZ;
  const nowM = minutesOf(zonedParts(now, tz));
  return (cfg.slots || []).filter((s) => {
    if (s.rail !== rail) return false;
    if (excludeId && s.id === excludeId) return false;
    return hhmmMinutes(s.hhmm) > nowM;
  });
}

export function reservedAhead(cfg, now, rail, excludeId) {
  return slotsAhead(cfg, now, rail, excludeId).reduce((a, s) => a + (s.cost || 1), 0);
}

export function estimatedSpentSlots(cfg, now, rail, graceMin = 20) {
  const tz = cfg.timezone || DEFAULT_TZ;
  const nowM = minutesOf(zonedParts(now, tz));
  let spent = 0;
  for (const s of cfg.slots || []) {
    if (s.rail !== rail) continue;
    if (nowM >= hhmmMinutes(s.hhmm) + graceMin) spent += s.cost || 1;
  }
  return spent;
}

export function decide(rail, opts = {}) {
  const cfg = opts.cfg;
  const spec = (cfg.rails || {})[rail] || {};
  const now = opts.now || new Date();
  const cost = opts.cost == null ? 1 : opts.cost;
  const isPeak = !!opts.isPeak;
  const isP0 = !!opts.isP0;
  const slotId = opts.slotId || null;
  const layer = spec.layer || "";
  const kind = spec.kind || "rate";
  const tz = spec.renewal_tz || cfg.timezone || DEFAULT_TZ;
  const renewal = spec.renewal_hhmm || "00:00";

  const base = (decision, remaining, hoursLeft, hourlyCap, spendable, reserved, reason) => ({
    decision,
    rail,
    remaining,
    hours_left: hoursLeft,
    hourly_cap: hourlyCap,
    spendable,
    reserved,
    cost,
    reason,
    layer,
    is_peak: isPeak,
    is_p0: isP0,
  });

  if (spec.hard_cap === 0) {
    return base(STASIS, 0, 0, 0, 0, 0, spec.note || "rail forbidden");
  }
  if (kind === "hard") {
    const cap = spec.hard_cap || spec.limit || 0;
    const rem = opts.remaining == null ? cap : opts.remaining;
    if (rem <= 0) return base(STASIS, rem, 0, 0, 0, 0, `hard cap ${cap} reached`);
    return base(ALLOW, rem, 0, 0, rem, 0, `hard cap ${cap}, remaining ${rem}`);
  }

  let hoursLeft;
  if (opts.resetUnix != null) hoursLeft = hoursUntilUnix(now, opts.resetUnix);
  else if (spec.window === "rolling_hour") hoursLeft = 1;
  else if (spec.window_sec) hoursLeft = Math.max(spec.window_sec / 3600, 1 / 60);
  else hoursLeft = hoursUntilRenewal(now, renewal, tz);
  let remaining = opts.remaining;
  if (remaining == null) {
    remaining = kind === "slots" && spec.daily_limit != null
      ? Math.max(0, spec.daily_limit - estimatedSpentSlots(cfg, now, rail))
      : spec.daily_limit || spec.limit || 0;
  }
  const hourlyCap = hoursLeft ? remaining / hoursLeft : remaining;
  let reserved = 0;
  if (kind === "slots") {
    reserved = reservedAhead(cfg, now, rail, isPeak || slotId ? slotId : null);
  }
  const spendable = remaining - reserved;

  if (remaining <= 0 && !isP0) return base(STASIS, remaining, hoursLeft, hourlyCap, spendable, reserved, "quota exhausted until renewal");
  if (remaining <= 0 && isP0) return base(P0_BORROW, remaining, hoursLeft, hourlyCap, spendable, reserved, "P0 borrows past empty pool — still no retry-loop");

  if (kind === "slots") {
    if (isP0) return base(remaining >= cost ? ALLOW : P0_BORROW, remaining, hoursLeft, hourlyCap, spendable, reserved, "P0 spends even if it trims a later peak");
    if (isPeak || slotId) {
      if (remaining >= cost) return base(ALLOW, remaining, hoursLeft, hourlyCap, spendable, reserved, "budgeted slot / reserved peak");
      return base(STASIS, remaining, hoursLeft, hourlyCap, spendable, reserved, "peak has no remaining");
    }
    if (spendable < cost) return base(HOLD, remaining, hoursLeft, hourlyCap, spendable, reserved, `protect ${reserved} reserved slot(s) still ahead`);
    return base(ALLOW, remaining, hoursLeft, hourlyCap, spendable, reserved, "unscheduled spend within contingency");
  }

  if (remaining < cost) return base(STASIS, remaining, hoursLeft, hourlyCap, remaining, 0, "remaining < cost");
  if (cost > hourlyCap + 1e-9) return base(HOLD, remaining, hoursLeft, hourlyCap, remaining, 0, `cost ${cost} > hourly_cap ${hourlyCap.toFixed(4)} (pace until renewal)`);
  return base(ALLOW, remaining, hoursLeft, hourlyCap, remaining, 0, "within hourly average");
}

export function monitorIntervalSec(now, cfg) {
  const spec = (cfg.rails || {})["local-monitor"] || {};
  const night = spec.night_interval_sec || 900;
  const day = spec.day_interval_sec || 300;
  return isNight(now, spec, cfg.timezone) ? night : day;
}

export function snapshot(cfg, now = new Date(), live = {}) {
  const tz = cfg.timezone || DEFAULT_TZ;
  const parts = zonedParts(now, tz);
  const railsOut = {};
  for (const name of Object.keys(cfg.rails || {})) {
    const extra = live[name] || {};
    const v = decide(name, { cfg, now, remaining: extra.remaining, resetUnix: extra.reset_unix });
    const spec = cfg.rails[name];
    railsOut[name] = {
      ...v,
      spec_note: spec.note,
      kind: spec.kind,
      unit: spec.unit,
      daily_limit: spec.daily_limit || spec.limit || spec.hard_cap,
    };
  }
  const slots = (cfg.slots || []).map((s) => {
    const v = decide(s.rail, {
      cfg,
      now,
      isPeak: !!s.peak,
      slotId: s.id,
      remaining: railsOut[s.rail]?.remaining,
    });
    return { ...s, verdict: v.decision, reason: v.reason };
  });
  return {
    schema: "stratamesh.metabolism.v1",
    at: now.toISOString(),
    lisbon: `${parts.year}-${pad(parts.month)}-${pad(parts.day)}T${pad(parts.hour)}:${pad(parts.minute)}:${pad(parts.second)}`,
    hour_lisbon: parts.hour,
    night: isNight(now, (cfg.rails || {})["local-monitor"], tz),
    monitor_interval_sec: monitorIntervalSec(now, cfg),
    cf_cron_hard_cap: cfg.cf_cron_hard_cap,
    formula: cfg.formula,
    rails: railsOut,
    slots,
    lab_honest: true,
    no_sixth_cron: true,
  };
}

File: ops/lib/metabolism.py
#!/usr/bin/env python3
"""Metabolic stasis — remaining / hours_until_renewal, all rails.

Lab · Node · StrataMesh share one circadian budget. Discrete Grok fires use
slot reservation (peaks 09:00 / 18:00 / 23:00 Lisbon). Continuous APIs use
hourly_cap = remaining / hours_left. Never add a 6th Cloudflare cron.
"""
from __future__ import annotations

import json
import math
from dataclasses import dataclass, asdict
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any, Optional
from zoneinfo import ZoneInfo

ROOT = Path(__file__).resolve().parents[1]
RAILS_PATH = ROOT / "config" / "rails.json"

ALLOW = "ALLOW"
HOLD = "HOLD"
STASIS = "STASIS"
P0_BORROW = "P0_BORROW"

LISBON = ZoneInfo("Europe/Lisbon")


def load_rails(path: Path = RAILS_PATH) -> dict:
    return json.loads(path.read_text())


def parse_hhmm(hhmm: str) -> tuple[int, int]:
    h, m = hhmm.split(":")
    return int(h), int(m)


def as_tz(now: Optional[datetime], tz_name: str) -> datetime:
    tz = ZoneInfo(tz_name)
    if now is None:
        now = datetime.now(timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)
    return now.astimezone(tz)


def hours_until_renewal(
    now: Optional[datetime] = None,
    renewal_hhmm: str = "00:00",
    tz_name: str = "Europe/Lisbon",
) -> float:
    """Hours until the next renewal instant in tz_name (minimum 1 minute)."""
    local = as_tz(now, tz_name)
    rh, rm = parse_hhmm(renewal_hhmm)
    candidate = local.replace(hour=rh, minute=rm, second=0, microsecond=0)
    if candidate <= local:
        candidate = candidate + timedelta(days=1)
    hours = (candidate - local).total_seconds() / 3600.0
    return max(hours, 1.0 / 60.0)


def hours_until_unix(now: Optional[datetime], reset_unix: int) -> float:
    if now is None:
        now = datetime.now(timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)
    reset = datetime.fromtimestamp(reset_unix, tz=timezone.utc)
    hours = (reset - now.astimezone(timezone.utc)).total_seconds() / 3600.0
    return max(hours, 1.0 / 60.0)


def minutes_of_day(local: datetime) -> int:
    return local.hour * 60 + local.minute


def hhmm_minutes(hhmm: str) -> int:
    h, m = parse_hhmm(hhmm)
    return h * 60 + m


def is_night(now: Optional[datetime] = None, spec: Optional[dict] = None) -> bool:
    spec = spec or {}
    tz = spec.get("renewal_tz") or "Europe/Lisbon"
    local = as_tz(now, tz)
    start = hhmm_minutes(spec.get("night_start") or "22:00")
    end = hhmm_minutes(spec.get("night_end") or "08:00")
    m = minutes_of_day(local)
    if start > end:
        return m >= start or m < end
    return start <= m < end


def slots_ahead(cfg: dict, now: Optional[datetime], rail: str, exclude_id: Optional[str] = None) -> list[dict]:
    local = as_tz(now, cfg.get("timezone") or "Europe/Lisbon")
    now_m = minutes_of_day(local)
    out = []
    for s in cfg.get("slots") or []:
        if s.get("rail") != rail:
            continue
        if exclude_id and s.get("id") == exclude_id:
            continue
        if hhmm_minutes(s["hhmm"]) > now_m:
            out.append(s)
    return out


def reserved_ahead(cfg: dict, now: Optional[datetime], rail: str, exclude_id: Optional[str] = None) -> float:
    return float(sum(s.get("cost") or 1 for s in slots_ahead(cfg, now, rail, exclude_id)))


def estimated_spent_slots(cfg: dict, now: Optional[datetime], rail: str, grace_min: int = 20) -> float:
    """Assume a budgeted slot has fired once we are grace_min past its hhmm."""
    local = as_tz(now, cfg.get("timezone") or "Europe/Lisbon")
    now_m = minutes_of_day(local)
    spent = 0.0
    for s in cfg.get("slots") or []:
        if s.get("rail") != rail:
            continue
        if now_m >= hhmm_minutes(s["hhmm"]) + grace_min:
            spent += float(s.get("cost") or 1)
    return spent


@dataclass
class Verdict:
    decision: str
    rail: str
    remaining: float
    hours_left: float
    hourly_cap: float
    spendable: float
    reserved: float
    cost: float
    reason: str
    layer: str = ""
    is_peak: bool = False
    is_p0: bool = False

    def as_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["hourly_cap"] = round(self.hourly_cap, 4)
        d["hours_left"] = round(self.hours_left, 4)
        d["spendable"] = round(self.spendable, 4)
        return d


def decide(
    rail: str,
    remaining: Optional[float] = None,
    now: Optional[datetime] = None,
    cost: float = 1.0,
    is_peak: bool = False,
    is_p0: bool = False,
    slot_id: Optional[str] = None,
    reset_unix: Optional[int] = None,
    cfg: Optional[dict] = None,
) -> Verdict:
    cfg = cfg or load_rails()
    spec = (cfg.get("rails") or {}).get(rail) or {}
    layer = spec.get("layer") or ""
    kind = spec.get("kind") or "rate"
    tz = spec.get("renewal_tz") or cfg.get("timezone") or "Europe/Lisbon"
    renewal = spec.get("renewal_hhmm") or "00:00"

    if spec.get("hard_cap") == 0:
        return Verdict(
            STASIS, rail, 0, 0, 0, 0, 0, cost,
            spec.get("note") or "rail forbidden",
            layer=layer, is_peak=is_peak, is_p0=is_p0,
        )

    if kind == "hard":
        cap = float(spec.get("hard_cap") or spec.get("limit") or 0)
        used = float(spec.get("used") or 0)
        rem = cap - used if remaining is None else float(remaining)
        if rem <= 0:
            return Verdict(STASIS, rail, rem, 0, 0, 0, 0, cost,
                           f"hard cap {cap} reached", layer=layer)
        return Verdict(ALLOW, rail, rem, 0, 0, rem, 0, cost,
                       f"hard cap {cap}, remaining {rem}", layer=layer)

    if reset_unix is not None:
        hours_left = hours_until_unix(now, reset_unix)
    elif spec.get("window") == "rolling_hour":
        hours_left = 1.0
    elif spec.get("window_sec"):
        hours_left = max(float(spec["window_sec"]) / 3600.0, 1.0 / 60.0)
    else:
        hours_left = hours_until_renewal(now, renewal, tz)

    daily = spec.get("daily_limit")
    if remaining is None:
        if kind == "slots" and daily is not None:
            remaining = max(0.0, float(daily) - estimated_spent_slots(cfg, now, rail))
        else:
            remaining = float(daily or spec.get("limit") or 0)

    remaining = float(remaining)
    hourly_cap = remaining / hours_left if hours_left else remaining

    reserved = 0.0
    if kind == "slots":
        reserved = reserved_ahead(cfg, now, rail, exclude_id=slot_id if (is_peak or slot_id) else None)
        if is_peak or slot_id:
            reserved = reserved_ahead(cfg, now, rail, exclude_id=slot_id)
    spendable = remaining - reserved

    if remaining <= 0 and not is_p0:
        return Verdict(STASIS, rail, remaining, hours_left, hourly_cap, spendable, reserved, cost,
                       "quota exhausted until renewal", layer=layer, is_peak=is_peak, is_p0=is_p0)
    if remaining <= 0 and is_p0:
        return Verdict(P0_BORROW, rail, remaining, hours_left, hourly_cap, spendable, reserved, cost,
                       "P0 borrows past empty pool — still no retry-loop", layer=layer, is_peak=True, is_p0=True)

    if kind == "slots":
        if is_p0:
            return Verdict(ALLOW if remaining >= cost else P0_BORROW, rail, remaining, hours_left,
                           hourly_cap, spendable, reserved, cost,
                           "P0 spends even if it trims a later peak", layer=layer, is_peak=is_peak, is_p0=True)
        if is_peak or slot_id:
            if remaining >= cost:
                return Verdict(ALLOW, rail, remaining, hours_left, hourly_cap, spendable, reserved, cost,
                               "budgeted slot / reserved peak", layer=layer, is_peak=True)
            return Verdict(STASIS, rail, remaining, hours_left, hourly_cap, spendable, reserved, cost,
                           "peak has no remaining", layer=layer, is_peak=True)
        if spendable < cost:
            return Verdict(HOLD, rail, remaining, hours_left, hourly_cap, spendable, reserved, cost,
                           f"protect {reserved} reserved slot(s) still ahead", layer=layer)
        return Verdict(ALLOW, rail, remaining, hours_left, hourly_cap, spendable, reserved, cost,
                       "unscheduled spend within contingency", layer=layer)

    # rate rails: this hour may spend at most hourly_cap
    if remaining < cost:
        return Verdict(STASIS, rail, remaining, hours_left, hourly_cap, remaining, 0, cost,
                       "remaining < cost", layer=layer)
    if cost > hourly_cap + 1e-9:
        return Verdict(HOLD, rail, remaining, hours_left, hourly_cap, remaining, 0, cost,
                       f"cost {cost} > hourly_cap {hourly_cap:.4f} (pace until renewal)", layer=layer)
    return Verdict(ALLOW, rail, remaining, hours_left, hourly_cap, remaining, 0, cost,
                   "within hourly average", layer=layer)


def monitor_interval_sec(now: Optional[datetime] = None, cfg: Optional[dict] = None) -> int:
    cfg = cfg or load_rails()
    spec = (cfg.get("rails") or {}).get("local-monitor") or {}
    night = int(spec.get("night_interval_sec") or 900)
    day = int(spec.get("day_interval_sec") or 300)
    return night if is_night(now, spec) else day


def snapshot(now: Optional[datetime] = None, live: Optional[dict] = None, cfg: Optional[dict] = None) -> dict:
    """Full metabolic picture for dashboard / handoff."""
    cfg = cfg or load_rails()
    live = live or {}
    now = now or datetime.now(timezone.utc)
    lisbon = as_tz(now, cfg.get("timezone") or "Europe/Lisbon")
    rails_out = {}
    for name, spec in (cfg.get("rails") or {}).items():
        kwargs: dict[str, Any] = {"cfg": cfg, "now": now}
        if name in live:
            if "remaining" in live[name]:
                kwargs["remaining"] = live[name]["remaining"]
            if "reset_unix" in live[name]:
                kwargs["reset_unix"] = live[name]["reset_unix"]
        v = decide(name, **kwargs)
        rails_out[name] = {**v.as_dict(), "spec_note": spec.get("note"), "kind": spec.get("kind"),
                           "unit": spec.get("unit"), "daily_limit": spec.get("daily_limit") or spec.get("limit") or spec.get("hard_cap")}
    slots = []
    for s in cfg.get("slots") or []:
        v = decide(s["rail"], now=now, is_peak=bool(s.get("peak")), slot_id=s["id"], cfg=cfg,
                   remaining=rails_out.get(s["rail"], {}).get("remaining"))
        slots.append({**s, "verdict": v.decision, "reason": v.reason})
    return {
        "schema": "stratamesh.metabolism.v1",
        "at": now.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "lisbon": lisbon.strftime("%Y-%m-%dT%H:%M:%S%z"),
        "hour_lisbon": lisbon.hour,
        "night": is_night(now, (cfg.get("rails") or {}).get("local-monitor")),
        "monitor_interval_sec": monitor_interval_sec(now, cfg),
        "cf_cron_hard_cap": cfg.get("cf_cron_hard_cap"),
        "formula": cfg.get("formula"),
        "rails": rails_out,
        "slots": slots,
        "lab_honest": True,
        "no_sixth_cron": True,
    }


def main(argv: Optional[list[str]] = None) -> int:
    import sys
    argv = argv if argv is not None else sys.argv[1:]
    snap = snapshot()
    if argv and argv[0] == "--pretty":
        print(json.dumps(snap, indent=2))
    else:
        print(json.dumps(snap, separators=(",", ":")))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
